### ISSUE TITLE
The first pixel of a flat is not very representative

### ISSUE THREAD

**Issue by RamonUnch** [viti95/FastDoom#25]
I was playing with -flatsurfaces and I realized that teleporters pads were brown and not red.
From reading the source it seems that the R_DrawPlanesFlatSurfaces*() procs use the first pixel of the flat as well as the R_DrawSpanFlat function.
I tried with an offet somewhere in the middle of 4096 ie:1993 and I think it gives beter results, see below.
Maybe someone would like to look through all the flats in doom and decide which is the best offset to have the most representative color out of each flat. 
Original FDoom flat reference offset = 0
![FDoom_flat_index0](https://user-images.githubusercontent.com/74856804/110233972-af7f4780-7f1f-11eb-80c7-ddbd5c3af271.png)

Flat reference offset = 1993
![FDoom_flat_index1993](https://user-images.githubusercontent.com/74856804/110233985-c9b92580-7f1f-11eb-81e5-02d4b50d8560.png)

**Comment by AXDOOMER** [comment]
Doing an average of all the pixels on the flat texture could be a better idea.

**Comment by viti95** [comment]
You're right @AXDOOMER (for example OptiDoom uses this idea), but for now it's a good compromise between speed and visual quality. I've tested different values and for now 1850 makes it even better. Thank's @RamonUnch !!

**Comment by RamonUnch** [comment]
I will try again with 1850. I was also thinking to make the average color but it would be time consuming. the ideal solution would be to precalculate a mapping table flat<->color and load it with fdoom. this table could be even fine-tuned by hand to give the best results.

**Comment by sandro-tishin** [comment]
> You're right [@AXDOOMER](https://github.com/AXDOOMER) (for example OptiDoom uses this idea), but for now it's a good compromise between speed and visual quality. I've tested different values and for now 1850 makes it even better. Thank's [@RamonUnch](https://github.com/RamonUnch) !!

Have been testing some possible implementations of deducing the best approximation of flat color. As it seems, taking average color and mapping it back to the Doom palette causes some strange results like reddish surfaces turning pink.  On the other hand, the method of finding the most common color index works surprisingly well for vanilla textures; likely reason is the use of very few color ramps (or even one) per texture. This naturally leads to selection of the most prominent color from one of ramps.
Loading time is not a problem here; collecting a frequency histogram should run for about 1ms per texture even on 386, so calculation time will be dwarfed by the disk I/O. And this cost is paid only once in the game preloading phase.

I can provide a complete patch if needed. Most of work is done, just needs some cleanup.

**Comment by viti95** [comment]
Oh, this is great! I like the most common color approach, so go ahead and send the patch.

By the way, I've been meaning to do a memory optimization for this. We don't actually need to keep the whole texture in RAM once the histogram is done.

**Comment by sandro-tishin** [comment]
Fine, will do that in a day or two. The memory optimization is already done. Flat colors are stored in byte array covering all the flat index range, and texture is discarded after finding the representative color. This does not interfere with use of flats in UI backgrounds (intermission screen, etc.), because UI reads them directly, bypassing all level loading and management logic.
It is possible to extend implementation to choosing color for flat rendering of walls, but this is definitely a separate issue. Masked and composite textures are not that simple.
BTW, have been playtesting for a some time, the only problematic texture so far is the bottom of rift at the Industrial Zone (MAP15); selected color is light brown. Not sure if averaging texel colors would help anyway, because only minor area of that flat is red.
